Add per-bone meta to Skeleton3D

demolke opened the pull request:
- Bones are not represented as `Node`s and don't inherit get/set meta functionality from `Object`, so the skeleton has to carry the meta information similarly to how other per-bone properties are handled.
- Adds support for GLTF import/export

This was split off of https://github.com/godotengine/godot/pull/86183 which only deals with import/export of metadata for nodes, meshes and materials - all of which do inherit from Object and therefore support metadata already.

Supporting the same for bones requires changes to `Skeleton3D` datastructure, so it's better to have the discussion separate.

![image](https://github.com/user-attachments/assets/b1d0749a-5493-4560-953b-d293bab85fd3)

demolke commented:
I would like to get https://github.com/godotengine/godot/pull/86183 merged first, so that this can be rebased and will be only one commit. Did not figure out how to tell github that this PRs diffbase is https://github.com/godotengine/godot/pull/86183

AThousandShips commented:
There's no such system AFAIK, just make it clear in the PR description that it depends on that PR and the production team and reviewers will keep that in mind

lyuma commented:
I like this idea because it enables round-tripping of GLTF extras.

Because we have heavily modified the skeleton class late in the release cycle already, I'm a bit hesitant to add another big change to the skeleton (adding another property per-bone). I think @TokageItLab would need to review the skeleton changes to make sure both the serialization and the API changes ok, but I don't want to put more on the plate for 4.3 this late in the cycle.

I'd feel more comfortable trying the original PR #86183 in for this release cycle, understanding that we won't have full round tripping of skeleton bone extras for now... and depending on feedback and when we have time, we can also add the skeleton bone support in the future.

That said, the code looks reasonably simple if I [read only the second commit](https://github.com/godotengine/godot/pull/87150/commits/8db507230f69556e7ec555639fae6909d6081917) (since it's stacked on the other PR)... if Tokage approves it and we think the timing works, I think it's ok.

demolke committed:
Add per-bone meta to Skeleton3D

Individual bones are not represented as `Node`s in Godot, in order to support meta functionality for them the skeleton has to carry the information similarly to how other per-bone properties are handled.
- Also adds support for GLTF import/export

demolke commented:
I think this is now ready for review - majority of the PR is adding the bone metadata functionality to `Skeleton3D` while the import/export functionality is just three lines each.

I wanted to make "extras" into a constant that could be accessed from both C++ as well as GDScript, but haven't found documentation on how to do it - I saw in ClassDB there is only `bind_integer_constant`, so I guess String/char* constants are not supported.

TokageItLab commented:
I wonder if this approach would cause the bone meta to remain in Skeleton when a bone is renamed or removed.

If we want bones to have meta, I think we should put it in Bone, but since there may be other places where we want to have individual meta, so I think we should have a base class with meta instead of strict. Is it time to define a weaker base class than Object? Since Object class seems to be a bit exaggerated.

demolke commented:
True, putting it on Bone makes it much cleaner, done and thanks for the pointer.

> Is it time to define a weaker base class than Object?
As far as I understand you can't access Bone from GDScript, so if we were to introduce a superclass that handles the meta (something like MetaHolder that Object and Bone would inherit from), the design wouldn't be too much cleaner - we would still have to have the trampoline in Skeleton3D that redirects the meta requests to Bone.

TokageItLab reviewed (approved):
I think the structure on the Skeleton3D side is fine. Need consensus from the asset-pipeline team on the glTFdocument and serialization.

AThousandShips commented on doc/classes/Skeleton3D.xml:
```diff
@@ -171,6 +186,14 @@
+			<param index="0" name="bone_idx" type="int" />
+			<param index="1" name="key" type="StringName" />
+			<description>
+				Returns if bone metadata for [param bone_idx] with key [param key] exists.
```
```suggestion
				Returns whether there exists any bone metadata for [param bone_idx] with key [param key].
```

AThousandShips commented on scene/3d/skeleton_3d.cpp:
```diff
@@ -531,6 +540,50 @@ void Skeleton3D::set_bone_name(int p_bone, const String &p_name) {
 
+Variant Skeleton3D::get_bone_meta(int p_bone, const StringName &p_key) const {
+	const int bone_size = bones.size();
+	ERR_FAIL_INDEX_V(p_bone, bone_size, "");
```
```suggestion
	ERR_FAIL_INDEX_V(p_bone, bone_size, Variant());
```
Otherwise it'll return `Variant()` if there's a bone but it doesn't have a mega with the key, and `""` if there's no such bone

AThousandShips commented on scene/3d/skeleton_3d.cpp:
```diff
@@ -531,6 +540,50 @@ void Skeleton3D::set_bone_name(int p_bone, const String &p_name) {
+
+TypedArray<StringName> Skeleton3D::_get_bone_meta_list_bind(int p_bone) const {
+	const int bone_size = bones.size();
+	ERR_FAIL_INDEX_V(p_bone, bone_size, {});
```
```suggestion
	ERR_FAIL_INDEX_V(p_bone, bone_size, TypedArray<StringName>());
```
Be explicit about return values

AThousandShips commented on doc/classes/Skeleton3D.xml:
```diff
@@ -99,6 +99,21 @@
+			<return type="StringName[]" />
+			<param index="0" name="bone_idx" type="int" />
+			<description>
+				Returns list of all metadata keys for [param bone_idx].
```
```suggestion
				Returns a list of all metadata keys for [param bone_idx].
```

AThousandShips commented:
You can remove the co-author as it's just from review

demolke commented:
Actually playing with this more this will need some more work, I'll switch it to draft.

AThousandShips commented:
Please set up format checks on your computer, see [here](https://docs.godotengine.org/en/stable/contributing/development/code_style_guidelines.html#using-clang-format-locally) instead of using the CI to do style checks for you, it taxes the system unnecessarily

demolke commented:
I've uploaded a new version which fixes deleting of meta and also contains the UI portion of the change. I've separated the "Add Metadata" dialog logic from Node into a separate Dialog and linked it to both node meta and bone meta.

![image](https://github.com/user-attachments/assets/b71384dd-5a15-451d-a721-1f1cbd342d93)

AThousandShips commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -183,16 +269,17 @@ void BoneTransformEditor::_property_keyed(const String &p_path, bool p_advance)
 	}
 	int selected = Skeleton3DEditor::get_singleton()->get_selected_bone();
 	List<PropertyInfo> props;
+	HashMap<StringName, bool> meta_seen;
```
```suggestion
	HashSet<StringName> meta_seen;
```
No need for a map where you never use the value

demolke replied:
Thanks, that's better

AThousandShips commented on scene/3d/skeleton_3d.cpp:
```diff
@@ -103,6 +103,8 @@ bool Skeleton3D::_set(const StringName &p_path, const Variant &p_value) {
 		set_bone_pose_rotation(which, p_value);
 	} else if (what == "scale") {
 		set_bone_pose_scale(which, p_value);
+	} else if (what == "bonemeta") {
```
```suggestion
	} else if (what == "bone_meta") {
```
Since the methods use `bone_meta` I'd say to use it everywhere

demolke replied:
Done

AThousandShips commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -224,9 +311,35 @@ void BoneTransformEditor::_update_properties() {
 	}
+	// UI for any bone metadata prop not seen during the iteration has to be deleted
+	for (KeyValue<StringName, EditorProperty *> iter : meta_editors) {
+		if (!meta_seen.find(iter.key)) {
```
```suggestion
		if (!meta_seen.has(iter.key)) {
```

demolke replied:
Done

AThousandShips reviewed (approved):
LGTM

akien-mga commented:
Thanks!
